### DISCUSSION TITLE
Refreshing session

### DISCUSSION THREAD

**Discussion by schreddies** [ory/kratos#1075] [Ideas]
Hi Folks!
I would like to start discussion about making session refresh for Kratos more useful.
## Current flow
Right now, according to documentation and my own experiments the refresh flow goes like this: user make request through API login for flow id, with flow id  tries to login with credentials (/password) and then getting session token ends login flow. Next, she/he calls the flow endpoint with `/?refresh=true` with `Authorization` header including `session_token` and the new `flow_id` is created. Then it can be used as the id flow for next login call with credentials.
## Expected
There is an endpoint, might be admin, when the similar call as for initiation login flow with `?refresh=true` is performed including the authenticated user (via `Authorization` header) the session_token will be prolonged or new token with new expiration value will be issue.

**Comment by schreddies** [comment]
@zepatrik @vinckr  FYI

**Reply by aeneasr** [reply]
Thank you for the write up!

The idea of that flow is to certify (authenticate) that a user is still in front of his/her device. This only works by requesting the credentials, a security key, ... only the user has access to. See: https://docs.github.com/en/github/authenticating-to-github/sudo-mode

It would be counterintuitive to change the `authenticated_at` of the session without re-confirming that the user is still there, as that would essentially contrary to the very basis of what authentication is.

We should however improve the wording in the docs on that!

**Reply by vinckr** [reply]
That makes sense, I will think of something to improve the wording there.

**Reply by schreddies** [reply]
@aeneasr I think that it makes sense to be able to prolong the authentication session without user interaction:
* better user experience,
* in some cases, even better security (less likely for shoulder surfer to get password/code right)
* shorter `session_token` duration
It could be like that, if user is currently active, through some activity (javascript `onfocus` attribute) additional request can be done - authenticated user can request to prolong the session. More constraints can be applied, like only 3 session prolong can be done. I can agree that basically we cannot say, that still that user is using this application, but for soma cases it would be beneficial to have slightly shorter token lifetime and that kind of 'refresh' than issuing long-life tokens - mobile apps for example. Maybe additional limit can be applied that every single n sec re-authentication needs to be applied.

**Reply by schreddies** [reply]
Ofc I definitely agree on that, that it cannot be implement to be infinite session!

**Reply by aeneasr** [reply]
I don't understand the security model behind that though - either the user is authenticated, with the appropriate factors like a password, hardware key, biometrics - or not. Using onFocus or similar things is as secure as having a longer session, because it does not add any verification nor security assurance other than a request being made from some internet participant.

What is the problem you are trying to solve - in terms of user flow?

**Reply by zepatrik** [reply]
One use case I remember we discussed already somewhere:
Consider an app that has a very short session lifespan (like 10min). This session is renewed as long as you are active. In this case you can assume that the user did not leave the device in between the requests because it is a very short period. E.g. my bank does this.
Whether this is secure is something different, but I guess that is a somehow valid use case.

**Reply by schreddies** [reply]
@aeneasr The security model behind that is similar to this form @zepatrik - if user is active, especially on mobile device (with other factors forced, like pin or biometry to unlock) we can prolong the session assuming that session is still active and user is still using the application. With requirement for session to be as short as possible we cannot force user to re-authenticate every 5 minutes but still she or he should be able to finish the business flow. Security measurement for that is the maksimum refresh/prolong session i.e. 3 times. Using short lived sessions (required) I would rather not to ask user every 5 minutes to re-authenticate.  Theoretically, instead using long live tokens  I am giving user the short live tokens and if application (or rather user) is still using it, I would like to prolong them. It is countermeasure for user leaving device unlocked and with valid session in the app, but still getting business flow done when it takes 15 minutes.

**Reply by aeneasr** [reply]
I see, that use case makes sense. I was always under the impression that the site would invalidate the session after 15 minutes but it probably makes more sense to have the session valid for only 15 minutes and be able to add another 15 minutes to it. So this wouldn't change the authenticated_at time, for example, it would just push the expires_at back for another 15 minutes. Did I get that right?

**Reply by schreddies** [reply]
I mean, short live tokens make more sense, especially in mobile cases, when user can exit the app without logging out and session will be still active. 
@aeneasr Yes, you are right, basically it might prolong the `expires_at` time for `n` minutes, where `n` is the session time and with that user can use application using short-life tokens and still have same `session_token`.

**Reply by ghost** [reply]
@schreddies @zepatrik (Open Banking recommended flow for authentication). Sessions require a rolling timer in the front (usually web) detecting user gestures to determine to refresh or not the user session (usually at 3/4 of the maximum user authentication cookie or token is sent to be refreshed or reobtain a new one without using prompts) without re-entering credentials, otherwise it logs the user out. You can have a notification to keep the user logged in. On mobile, refresh tokens with PKCE for example stored in KeyStore Android or KeyChain iOS can be used to prevent full reauthentification, just local to access refresh tokens even after 2 weeks if the user has MFA. Correct me if i am wrong but Kratos /?refresh=true with session token does perform just that right ? Extend the user session ? I didn't quite catch why the new flow-id is created.

**Reply by aeneasr** [reply]
`refresh=true` will re-authenticate using a prompt so it requires e.g. presenting your hardware key, password, ...

**Comment by zepatrik** [comment]
Btw, this is tracked and discussed in https://github.com/ory/kratos/issues/615
